### PR TITLE
Bugfix: Allow Properties with Carets in MLIR syntax.

### PR DESCRIPTION
**Pull request by AtticusKuhn**
Allow properties with carets in LLVM.

This is a syntax which was supported by the MLIR specification
[Properties](https://mlir.llvm.org/docs/LangRef/#properties)

but was not supported by our code.

### CODE REVIEW

**Comment by AtticusKuhn** [comment]
This was done as pair programming with @alexkeizer .

File: SSA/Core/MLIRSyntax/GenericParser.lean [hunk 1]
```diff
@@ -692,15 +715,19 @@ macro "[mlir_attr_entry|" entry:mlir_attr_entry "]" : term => do
...
-| `([mlir_attr_dict| {  $attrEntries,* } ]) => do
+| `([mlir_attr_dict| $[<%$caretLeft]? {  $attrEntries,* } $[>%$caretRight]? ]) => do
+        match caretLeft, caretRight with
+        | none, some y => Macro.throwErrorAt y "Caret closed without being opened"
+        | some x, none => Macro.throwErrorAt x "Caret opened without being closed"
+        | _ , _ =>
+
         let attrsList <- attrEntries.getElems.toList.mapM (fun x => `([mlir_attr_entry| $x]))
         let attrsList <- quoteMList attrsList (<- `(MLIR.AST.AttrEntry _))
         `(AttrDict.mk $attrsList)
```
Comment (alexkeizer) [inline SSA/Core/MLIRSyntax/GenericParser.lean:730]: Consider removing the whitespace line, and indenting the three lines below it

File: SSA/Core/MLIRSyntax/GenericParser.lean [hunk 2]
```diff
@@ -478,20 +478,43 @@ macro_rules
...
 
-syntax  "{" ("^" ident ("(" sepBy(mlir_bb_operand, ",") ")")? ":")? mlir_ops "}" : mlir_region
+syntax mlir_suffix_id := num <|> ident
+syntax  "{" ("^" mlir_suffix_id ("(" sepBy(mlir_bb_operand, ",") ")")? ":")? mlir_ops "}" : mlir_region
 syntax "[mlir_region|" mlir_region "]": term
 
+def getSuffixId : TSyntax ``mlir_suffix_id → String
+  | `(mlir_suffix_id| $x:ident) => x.getId.toString
+  | `(mlir_suffix_id| $x:num) => toString (x.getNat)
+  | _ => "" -- Should never happen, since `mlir_suffix_id` is a closed syntax definition
```
Comment (alexkeizer) [inline SSA/Core/MLIRSyntax/GenericParser.lean:507]: Please write a docstring for `getSuffixId`, that explains what it does.

**Review by alexkeizer — COMMENTED** [review]
@AtticusKuhn next time please wait for a code review before merging. 
Even if we did write it together, when pairing (especially if we do so at 6pm ;) ) I might take some shortcuts that I'll expect you to fix/polish before we merge.

For now, I have some minor comments, please open a follow up PR that makes the requested changes.

**Comment by AtticusKuhn** [comment]
> @AtticusKuhn next time please wait for a code review before merging. Even if we did write it together, when pairing (especially if we do so at 6pm ;) ) I might take some shortcuts that I'll expect you to fix/polish before we merge.
> 
> For now, I have some minor comments, please open a follow up PR that makes the requested changes.

Sorry, that was my bad. I'll make the changes.

**Comment by alexkeizer** [comment]
No worries, nothing bad happened. Also, don't feel like you have to make the changes now, this can definitely wait until Monday

**Comment by AtticusKuhn** [comment]
I made the changes in
https://github.com/opencompl/lean-mlir/pull/466

**Pull request merged**